rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+  <tr>
+   <td><strong>Output[3]:</strong>
+<p>
+_ExecutionResult object for ExampleValidator_
+   </td>
+  </tr>
+</table>
+
+Note that the user may have forgotten to run InteractiveContext.run() on
+upstream components in the dependency graph. Instead of implicitly running these
+upstream components, we remind the user to explicitly run upstream notebook
+cells (with a readable error message). We think this explicit mode of component
+execution is more notebook-friendly, and is easy to use with common notebook
+actions such as “Run All”, “Run Cells Before”, and “Run Cells After”.
+
+### Export to a selected orchestration engine (v0)
+
+#### Filter out InteractiveContext() objects
+
+##### Option 1: Replace InteractiveContext instances with dummy versions.
+
+1.  Search for possible import alias, e.g. `from
+    tfx.orchestration.interactive.interactive_context import InteractiveContext
+    as FooBar`
+
+*   Search for all instances of string ".*InteractiveContext", or the alias name
+    if found from prior step.
+*   Replace each instance with `DummyInteractiveContext`, which inherits from
+    InteractiveContext and basically does nothing / returns empty
+    ExecutionResult on .run().
+
+    ```
+    class DummyInteractiveContext(InteractiveContext):
+    def run(self,
+      component: base_component.BaseComponent,
+      enable_cache: bool = True):
+    return None
+    ```
+
+    1.  This should cover the case where the class definition is aliased.
+
+        ```
+        aliased_class = interactive_context.InteractiveContext
+        sess = aliased_class()
+        ```
+
+    1.  This should cover subsequent aliases of InteractiveContext instances.
+
+        ```
+        a = InteractiveContext()
+        b = a
+        c = InteractiveContext()
+        d = c
+        ```
+
+Cons:
+
+*   DummyInteractiveContext is now present/clutters the production pipeline code
+    (it's a no-op so mainly affects readability, not execution).
+*   Down the line, converting back to a notebook (replacing
+    DummyInteractiveContext with InteractiveContext) could be fragile.
+
+##### Option 2: Ensure InteractiveContext only runs in notebook context. **[recommended]**
+
+*   If InteractiveContext is run outside of a notebook context, just log a
+    warning and return.
+*   Bi-directional import to notebook from pipeline would "just work".
+
+Cons:
+
+*   InteractiveContext is still present in the production pipeline as a no-op /
+    affects readability.
+*   Puts the burden on user to scrub out calls to InteractiveContext.
+
+##### Option 3: Mark lines/cells to be skipped during export.
+
+Add custom magic to mark lines/cells as skip_for_export, can also be used by the
+user to skip scratch work in cells.
+
+Example line magic:
+
+```
+%skip_for_export sess = InteractiveContext()
+...
+%skip_for_export sess.run(example_gen)
+```
+
+Example cell magic:
+
+```
+%%skip_for_export
+# Cell contains scratch work that doesn't need to be exported.
+...
```

Comment:
What is supposed to be exported then? The use case being described here is not clear to me.